File: packages/agents/src/providers/anthropic.ts
import Anthropic from "@anthropic-ai/sdk";
import type { AgentMessage, LLMProvider, LLMStopReason, LLMTurn, Tool } from "../types.js";

export interface AnthropicProviderOptions {
  apiKey?: string;
  /** Point at a custom endpoint (a self-hosted gateway, a Bedrock/Vertex proxy, ...) instead of Anthropic's default API. */
  baseURL?: string;
  model?: string;
  maxTokens?: number;
  /**
   * How many times the underlying @anthropic-ai/sdk client retries a single
   * call on a retriable error (429/5xx/timeout/connection error) with its
   * own exponential backoff, before chat()/chatStream() throw. The SDK
   * already defaults this to 2 — exposed here because that default was
   * previously invisible and unconfigurable from createAnthropicProvider().
   * For falling back to a *different* provider once this is exhausted, see
   * createFallbackProvider().
   */
  maxRetries?: number;
}

const DEFAULT_MODEL = "claude-sonnet-5";
const DEFAULT_MAX_TOKENS = 4096;

/**
 * The Messages API rejects any message whose content is an empty string or an
 * empty array ("all messages must have non-empty content"), and Agent.run()
 * can produce both: an assistant turn that carried neither text nor tool
 * calls, and a `{role:"assistant", text:""}` the responseSchema repair path
 * pushes directly. Such a message is dropped rather than filled in — a turn
 * with nothing in it carries no information, and inventing placeholder text
 * would attribute words to the model it never produced.
 *
 * The test keys on "no content at all", not "no text": an assistant turn with
 * tool calls and no narration is both legitimate and common, and dropping it
 * would break every tool-use loop. See REMEDIATION 3.6.
 */
function toAnthropicMessages(messages: AgentMessage[]): Anthropic.MessageParam[] {
  return messages.flatMap((message): Anthropic.MessageParam[] => {
    if (message.role === "user") {
      if (!message.text) return [];
      return [{ role: "user", content: message.text }];
    }
    if (message.role === "tool") {
      const result = message.toolResult;
      if (!result) throw new Error("AgentMessage with role 'tool' is missing toolResult");
      return [
        {
          role: "user",
          // JSON.stringify(undefined) is undefined, not a string — which would
          // be an empty content block, the very thing this function exists to
          // prevent. Normalize a missing output to JSON null instead.
          content: [{ type: "tool_result", tool_use_id: result.id, content: JSON.stringify(result.output ?? null) }],
        },
      ];
    }
    // role === "assistant"
    const content: Anthropic.ContentBlockParam[] = [];
    if (message.text) content.push({ type: "text", text: message.text });
    for (const call of message.toolCalls ?? []) {
      content.push({ type: "tool_use", id: call.id, name: call.name, input: call.input as Record<string, unknown> });
    }
    if (content.length === 0) return [];
    return [{ role: "assistant", content }];
  });
}

/**
 * `max_tokens` is required by this API and defaults to 4096 above, so hitting
 * the cap is routine rather than exotic — and until REMEDIATION 3.2 nothing
 * read the field that says it happened. "pause_turn" is a long-running
 * server-tool turn the caller is meant to continue, not an ending, so it maps
 * to "other" rather than "end": Agent treats it as a normal turn, which is
 * the existing behaviour, instead of asserting a completion the model never
 * signalled.
 */
function toStopReason(stopReason: string | null | undefined): LLMStopReason | undefined {
  switch (stopReason) {
    case "end_turn":
    case "stop_sequence":
      return "end";
    case "tool_use":
      return "tool_calls";
    case "max_tokens":
      return "length";
    case "refusal":
      return "refusal";
    case null:
    case undefined:
      return undefined;
    default:
      return "other";
  }
}

/**
 * Thin adapter over @anthropic-ai/sdk's Messages API tool-use loop. One of
 * two built-in LLMProvider implementations proving the interface isn't
 * secretly hardcoded to one vendor — Agent/Crew never reference this module.
 */
export function createAnthropicProvider(options: AnthropicProviderOptions = {}): LLMProvider {
  const client = new Anthropic({
    apiKey: options.apiKey ?? process.env.ANTHROPIC_API_KEY,
    baseURL: options.baseURL,
    maxRetries: options.maxRetries,
  });
  const model = options.model ?? DEFAULT_MODEL;
  const maxTokens = options.maxTokens ?? DEFAULT_MAX_TOKENS;

  return {
    name: "anthropic",
    async chat({ system, messages, tools }: { system?: string; messages: AgentMessage[]; tools: Tool[] }): Promise<LLMTurn> {
      const response = await client.messages.create({
        model,
        max_tokens: maxTokens,
        system,
        messages: toAnthropicMessages(messages),
        tools: tools.map((t) => ({
          name: t.name,
          description: t.description,
          input_schema: t.inputSchema as Anthropic.Tool.InputSchema,
        })),
      });

      const textBlocks = response.content.filter((b): b is Anthropic.TextBlock => b.type === "text");
      const toolUseBlocks = response.content.filter((b): b is Anthropic.ToolUseBlock => b.type === "tool_use");

      return {
        text: textBlocks.map((b) => b.text).join("\n") || undefined,
        toolCalls: toolUseBlocks.map((b) => ({ id: b.id, name: b.name, input: b.input })),
        stop: toolUseBlocks.length === 0,
        stopReason: toStopReason(response.stop_reason),
        usage: { inputTokens: response.usage.input_tokens, outputTokens: response.usage.output_tokens },
      };
    },

    async chatStream(
      { system, messages, tools }: { system?: string; messages: AgentMessage[]; tools: Tool[] },
      onText: (delta: string) => void,
    ): Promise<LLMTurn> {
      const stream = client.messages.stream({
        model,
        max_tokens: maxTokens,
        system,
        messages: toAnthropicMessages(messages),
        tools: tools.map((t) => ({
          name: t.name,
          description: t.description,
          input_schema: t.inputSchema as Anthropic.Tool.InputSchema,
        })),
      });
      stream.on("text", (textDelta) => onText(textDelta));

      const response = await stream.finalMessage();
      const textBlocks = response.content.filter((b): b is Anthropic.TextBlock => b.type === "text");
      const toolUseBlocks = response.content.filter((b): b is Anthropic.ToolUseBlock => b.type === "tool_use");

      return {
        text: textBlocks.map((b) => b.text).join("\n") || undefined,
        toolCalls: toolUseBlocks.map((b) => ({ id: b.id, name: b.name, input: b.input })),
        stop: toolUseBlocks.length === 0,
        stopReason: toStopReason(response.stop_reason),
        usage: { inputTokens: response.usage.input_tokens, outputTokens: response.usage.output_tokens },
      };
    },
  };
}

File: packages/agents-python/berth_agents/providers/anthropic.py
"""Thin adapter over the `anthropic` package's Messages API tool-use loop.
One built-in LLMProvider implementation, proving the interface isn't secretly
hardcoded to one vendor — Agent/Crew never reference this module. Mirrors
@berth/agents' providers/anthropic.ts, including chat_stream()/usage."""

from __future__ import annotations

import json
import os
from typing import Any, Callable

from anthropic import AsyncAnthropic

from ..types import AgentMessage, LLMTurn, Tool, ToolCall, Usage

DEFAULT_MODEL = "claude-sonnet-5"
DEFAULT_MAX_TOKENS = 4096


def _to_anthropic_messages(messages: list[AgentMessage]) -> list[dict[str, Any]]:
    result: list[dict[str, Any]] = []
    for message in messages:
        if message.role == "user":
            result.append({"role": "user", "content": message.text or ""})
        elif message.role == "tool":
            if message.tool_result is None:
                raise ValueError("AgentMessage with role 'tool' is missing tool_result")
            result.append(
                {
                    "role": "user",
                    "content": [
                        {
                            "type": "tool_result",
                            "tool_use_id": message.tool_result.id,
                            "content": json.dumps(message.tool_result.output),
                        }
                    ],
                }
            )
        else:  # role == "assistant"
            content: list[dict[str, Any]] = []
            if message.text:
                content.append({"type": "text", "text": message.text})
            for call in message.tool_calls or []:
                content.append({"type": "tool_use", "id": call.id, "name": call.name, "input": call.input})
            result.append({"role": "assistant", "content": content})
    return result


class _AnthropicProvider:
    name = "anthropic"

    def __init__(self, client: AsyncAnthropic, model: str, max_tokens: int) -> None:
        self._client = client
        self._model = model
        self._max_tokens = max_tokens

    async def chat(self, *, system: str | None, messages: list[AgentMessage], tools: list[Tool]) -> LLMTurn:
        kwargs: dict[str, Any] = {
            "model": self._model,
            "max_tokens": self._max_tokens,
            "messages": _to_anthropic_messages(messages),
            "tools": [{"name": t.name, "description": t.description, "input_schema": t.input_schema} for t in tools],
        }
        if system is not None:
            kwargs["system"] = system

        response = await self._client.messages.create(**kwargs)

        text_blocks = [block for block in response.content if block.type == "text"]
        tool_use_blocks = [block for block in response.content if block.type == "tool_use"]

        return LLMTurn(
            text="\n".join(block.text for block in text_blocks) or None,
            tool_calls=[ToolCall(id=block.id, name=block.name, input=block.input) for block in tool_use_blocks],
            stop=len(tool_use_blocks) == 0,
            usage=Usage(input_tokens=response.usage.input_tokens, output_tokens=response.usage.output_tokens),
        )

    async def chat_stream(
        self,
        *,
        system: str | None,
        messages: list[AgentMessage],
        tools: list[Tool],
        on_text: Callable[[str], None],
    ) -> LLMTurn:
        kwargs: dict[str, Any] = {
            "model": self._model,
            "max_tokens": self._max_tokens,
            "messages": _to_anthropic_messages(messages),
            "tools": [{"name": t.name, "description": t.description, "input_schema": t.input_schema} for t in tools],
        }
        if system is not None:
            kwargs["system"] = system

        async with self._client.messages.stream(**kwargs) as stream:
            async for text_delta in stream.text_stream:
                on_text(text_delta)
            response = await stream.get_final_message()

        text_blocks = [block for block in response.content if block.type == "text"]
        tool_use_blocks = [block for block in response.content if block.type == "tool_use"]

        return LLMTurn(
            text="\n".join(block.text for block in text_blocks) or None,
            tool_calls=[ToolCall(id=block.id, name=block.name, input=block.input) for block in tool_use_blocks],
            stop=len(tool_use_blocks) == 0,
            usage=Usage(input_tokens=response.usage.input_tokens, output_tokens=response.usage.output_tokens),
        )


def create_anthropic_provider(
    *,
    api_key: str | None = None,
    base_url: str | None = None,
    model: str | None = None,
    max_tokens: int | None = None,
    max_retries: int | None = None,
) -> _AnthropicProvider:
    client_kwargs: dict[str, Any] = {"api_key": api_key or os.environ.get("ANTHROPIC_API_KEY")}
    if base_url is not None:
        client_kwargs["base_url"] = base_url
    if max_retries is not None:
        client_kwargs["max_retries"] = max_retries

    client = AsyncAnthropic(**client_kwargs)
    return _AnthropicProvider(client, model or DEFAULT_MODEL, max_tokens or DEFAULT_MAX_TOKENS)
